### PR TITLE
fix(dts): array buffer types (ES2024,ES2017)

### LINKED ISSUE #54636: fix(dts): Update type of `ArrayBuffer`
# lib Update Request

The problem is that proper types are not available when `Resizable ArrayBuffer` need to be created by passing `options` in the constructor.

## Configuration Check

My compilation *target* is `ESnext` and my *lib* is `the default`.

## Missing / Incorrect Definition

**`ArrayBuffer`** Constructor signature.

## Sample Code

```TypeScript
const buff = new ArrayBuffer(10, {
  maxByteLength: 100,
});

console.debug(buff.resizable);
console.debug(buff.byteLength);

buff.resize(100);
console.debug(buff.byteLength);
```

## Documentation Link

You can check the signature over here.

[ArrayBuffer() constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer/ArrayBuffer)

### PR DESCRIPTION
**Pull request by santifoo**
This should be an updated, and hopefully correct fix from the stale PR here https://github.com/microsoft/TypeScript/pull/54637.

In summary this PR is an extension of the previous PR with several modifications (to comments, interfaces, references etc.) The previous PR included ArrayBuffer-related interfaces that are part of ES2024.

It passes most of the tests with the exceptions of baselines from what I understand, so those are accepted and send with a commit.

Would be useful to get some feedback if this looks good in principle @petamoriken @sandersn, I've included most of the reviews you made in the stale PR that I linked earlier. 

Fixes #54636

### CODE REVIEW

**Comment by santifoo** [comment]
I'm unsure what the licence implies about original code, part of this PR is based on a stale PR linked at the top, since it's impossible to rebase the original one (it's 800 commits behind.)

**Comment by jakebailey** [comment]
Not a lawyer, but they gave the PR away in https://github.com/microsoft/TypeScript/pull/54637#issuecomment-1859019824, so I would just add `Co-authored-by: ...` to your commit to credit the original code.

**Comment by santifoo** [comment]
> Not a lawyer, but they gave the PR away in https://github.com/microsoft/TypeScript/pull/54637#issuecomment-1859019824, so I would just add `Co-authored-by: ...` to your commit to credit the original code.

Thanks, sounds good. I can only do that revising and forcing a push I think. Is this fine for you?

**Comment by jakebailey** [comment]
Yes, feel free to force push

**Comment by santifoo** [comment]
> Yes, feel free to force push

done, it did work because the profile picture is correct in the co-authored commit.

**Comment by santifoo** [comment]
@microsoft-github-policy-service agree

**Comment by santifoo** [comment]
Still needs some tests https://github.com/microsoft/TypeScript/pull/54637#pullrequestreview-1504255858

and also for the new def of `slice()` and the empty ArrayBuffer and such constructors

**Comment by BlackAsLight** [comment]
Any update on this?

**Comment by santifoo** [comment]
> Any update on this? 

I can't spend the time now. I think it needs some tests only. Feel free to take over the PR.

File: src/lib/esnext.arraybuffer.d.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+     */
+    get resizable(): boolean;
+
+    /**
+     * Resizes the ArrayBuffer to the specified size (in bytes).
+     * @param newByteLength The new length, in bytes, to resize the ArrayBuffer to.
+     */
+    resize(newByteLength: number): undefined;
+}
+
```
Comment (fwqaaq) [inline src/lib/esnext.arraybuffer.d.ts:17]: I think there are other methods and properties that need to be added. Like this: <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer/detached>

**Comment by petamoriken** [comment]
Oh, It appears that the account has been lost. I'll take it instead.

**Comment by guest271314** [comment]
@petamoriken Status?

**Comment by petamoriken** [comment]
@guest271314 Please see #58573

**Pull request closed without merging**